tests/Doctrine/Tests/ORM/Functional/Ticket/GH7941Test.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use DateTimeImmutable;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use function ltrim;
+use function strlen;
+
+/** @group GH7941 */
+final class GH7941Test extends OrmFunctionalTestCase
+{
+    private const PRODUCTS = [
+        ['name' => 'Test 1', 'price' => '100', 'square_root' => '/^10(\.0)?$/'],
+        ['name' => 'Test 2', 'price' => '100', 'square_root' => '/^10(\.0)?$/'],
+        ['name' => 'Test 3', 'price' => '100', 'square_root' => '/^10(\.0)?$/'],
+        ['name' => 'Test 4', 'price' => '25', 'square_root' => '/^5(\.0)?$/'],
+        ['name' => 'Test 5', 'price' => '25', 'square_root' => '/^5(\.0)?$/'],
+        ['name' => 'Test 6', 'price' => '-25', 'square_root' => null],
+    ];
+
+    protected function setUp() : void
+    {
+        parent::setUp();
+
+        $this->setUpEntitySchema([GH7941Product::class]);
+
+        foreach (self::PRODUCTS as $product) {
+            $this->_em->persist(new GH7941Product($product['name'], $product['price']));
+        }
+
+        $this->_em->flush();
+        $this->_em->clear();
+    }
+
+    /** @test */
+    public function typesShouldBeConvertedForDQLFunctions() : void
+    {
+        $query = $this->_em->createQuery(
+            'SELECT
+                 COUNT(product.id) as count,
+                 SUM(product.price) as sales,
+                 AVG(product.price) as average
+             FROM ' . GH7941Product::class . ' product'
+        );
+
+        $result = $query->getSingleResult();
+
+        self::assertSame(6, $result['count']);
+        self::assertSame('325', $result['sales']);
+        self::assertRegExp('/^54\.16+7$/', $result['average']);
```

Comment:
@Grafikart this is what I meant with a more functional approach :smile: 

The main advantage is that we only rely on the public API here, which makes it easier to detect BC-breaks.

By using `self::assertSame()` we check that both value and type match expected value and `self::assertRegExp()` only accepts strings.